File: simple-java/src/graphs/DFS.java
package graphs;

import java.util.ArrayList;
import java.util.Arrays;

public class DFS {
    public static void main(String[] args) {
        var V = 5;
        var adj = new ArrayList<ArrayList<Integer>>(
                Arrays.asList(
                        new ArrayList<>(Arrays.asList(2, 3, 1)),
                        new ArrayList<>(Arrays.asList(0)),
                        new ArrayList<>(Arrays.asList(0, 4)),
                        new ArrayList<>(Arrays.asList(0)),
                        new ArrayList<>(Arrays.asList(2))));

        var dfs = new DFS();
        System.out.println(dfs.dfsOfGraph(V, adj));
    }

    private void dfs(int node, ArrayList<ArrayList<Integer>> adjList, boolean[] visited, ArrayList<Integer> ansList) {
        visited[node] = true;
        ansList.add(node);

        for (Integer adj : adjList.get(node)) {
            if (!visited[adj]) {
                dfs(adj, adjList, visited, ansList);
            }
        }
    }

    public ArrayList<Integer> dfsOfGraph(int V, ArrayList<ArrayList<Integer>> adj) {
        boolean[] visited = new boolean[V];
        ArrayList<Integer> ansList = new ArrayList<>();
        dfs(0, adj, visited, ansList);
        return ansList;
    }
}

File: simple-py/dfs/dfs.py
from typing import List


def dfsOfGraph(V, adj):

    def dfs(
        node: int, adjList: List[List[int]], visited: List[bool], ansList: List[int]
    ):
        visited[node] = True
        ansList.append(node)

        for adj in adjList[node]:
            if not visited[adj]:
                dfs(adj, adjList, visited, ansList)

    visited: List[bool] = [False for _ in range(V)]
    answerList = []

    for i in range(V):  # This for loop is in case of connected components
        if not visited[i]:
            dfs(i, adj, visited, answerList)

    return answerList


V = 5
adj = [[2, 3, 1], [0], [0, 4], [0], [2]]

print(dfsOfGraph(V, adj))
